core/src/test/java/kafka/test/api/ShareConsumerTest.java

```diff
@@ -1459,6 +1464,243 @@ public void testLsoMovementByRecordsDeletion() {
+        }
+        assertEquals(1, numInternalRecordsConsumed);
+
+        internalConsumer.close();
+        producer.close();
+    }
+
+    @Test
+    public void testReadShareGroupState() throws Exception {
+        Set<String> topics = listTopics().names().get();
+        // The __share_group_state topic is created on the first FIND_COORDINATOR RPC call for any share group topic partition.
+        // Since that does not happen automatically on cluster creation, we expect that the cluster will not have
+        // __share_group_state topic until yet.
+        assertFalse(topics.contains(SHARE_GROUP_STATE_TOPIC_NAME));
+
+        String key1 = "key1";
+        String key2 = "key2";
+        String value1 = "value1";
+        String value2 = "value2";
+
+        ProducerRecord<byte[], byte[]> record1 = new ProducerRecord<>(tp.topic(), 0, null, key1.getBytes(), value1.getBytes());
+        ProducerRecord<byte[], byte[]> record2 = new ProducerRecord<>(tp.topic(), 0, null, key2.getBytes(), value2.getBytes());
+
+        KafkaProducer<byte[], byte[]> producer = createProducer(new ByteArraySerializer(), new ByteArraySerializer());
+
+        // consumer to consume records produced to the custom topic
+        KafkaShareConsumer<byte[], byte[]> shareConsumer = createShareConsumer(new ByteArrayDeserializer(), new ByteArrayDeserializer(), "group1");
+        shareConsumer.subscribe(Collections.singleton(tp.topic()));
+
+        try {
+            producer.send(record1).get();
+        } catch (Exception e) {
+            fail("Failed to send records: " + e);
+        }
+
+        int numRecordsConsumed = 0;
+        ConsumerRecords<byte[], byte[]> consumedRecords = null;
+
+        while (numRecordsConsumed == 0) {
+            // First poll will initialise the share partition, which will internally try to read the state from the persister.
+            // Since this is the first persister related call, it will result in a lot of events happening in the background
+            // including creation of __share_group_state topic and loading of coordinator shards' in-memory states.
+            consumedRecords = shareConsumer.poll(Duration.ofMillis(2000));
+            numRecordsConsumed += consumedRecords.count();
+        }
+
+        assertEquals(1, numRecordsConsumed);
+        ConsumerRecord<byte[], byte[]> consumedRecord = consumedRecords.records(tp).iterator().next();
+        assertEquals(new String(consumedRecord.key()), key1);
+        assertEquals(new String(consumedRecord.value()), value1);
+
+        long duration = 1000; // 1 seconds
+
+        long startTime = System.currentTimeMillis();
+
+        // Poll for 1 second to ensure acknowledgement is sent for the fetched record
+        while (System.currentTimeMillis() - startTime < duration) {
+            shareConsumer.poll(Duration.ofMillis(2000));
+        }
+
+        topics = listTopics().names().get();
+        assertTrue(topics.contains(SHARE_GROUP_STATE_TOPIC_NAME));
```

Comment:
use wait for condition

Reply:
Thanks for the review. Actually at this point I ma not waiting for the creation of internal topic, that would have been created automatically before the consumer consumes its first message. the extra 1 second poll has been added to make sure any remaining acknowledgements for the consumed messages are sent to the broker.